### PR TITLE
Show a warning for missing attention masks when pad_token_id is not None

### LINKED ISSUE #16136: Add warning message if model uses `input_ids` that include padding tokens, but no `attention_mask` is provided.
## **First good issue** 

A current error is that a user forwards a batched tensor of `input_ids` that include a padding token, e.g. ```input_ids = torch.tensor([["hello", "this", "is", "a", "long", "string"], ["hello", "<pad>", "<pad>", "<pad>", "<pad>"]]```

In this case, the `attention_mask` should be provided as well. Otherwise the output hidden_states will be incorrectly computed. This is quite a common silent error IMO. 

With @LysandreJik @sgugger, we have decided to not **automatically** create the `attention_mask` that masks out the padding tokens in this case because of the reasons explains here: https://github.com/huggingface/transformers/issues/15479#issuecomment-1066639938 . However as pointed out in https://github.com/huggingface/transformers/issues/15479, we should IMO at least displa a warning since this error happens a lot IMO. 

As a first good issue, one could add such a warning to the BertModel in a first case which would go something like:

```py
if attention_mask is not None and (input_ids == pad_token_id).any():
    logger.warn("display nice warning here....")
```

What do you think @sgugger @LysandreJik ?

### PR DESCRIPTION
**Pull request by hackyon**
# What does this PR do?

Fixes #16136

Shows a one-time warning message when the pad_token_id is not None and no attention masks are given. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? #16136 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@gante @ydshieh

### CODE REVIEW

**Comment by hackyon** [comment]
@ydshieh @gante

I've added the warning message as per #21916 and also a short test. Please let me know if you have any suggestions. If it looks good, I can copy/paste the warning to more models as well. Thanks!

**Comment by gante** [comment]
@hackyon LGTM 👍 

But before you spend time propagating the change to the other models, let's also get a green light from a core maintainer (cc @sgugger )

**Review by sgugger — COMMENTED** [review]
Thanks for your PR but I'm not in favor of this. There are a lot of false positives this is going to generate: an attention mask should be passed if there are any tokens in the inputs that are the pad token. It's not necessary to pass one just because the model has a pad token ID.

**Comment by hackyon** [comment]
Thanks for the review. 

Would it be better to check for the presence of the pad_token_id inside input_ids first before throwing the error, as per
https://github.com/huggingface/transformers/pull/17444/files? If so, I can make the change to reflect that here.

**Comment by hackyon** [comment]
@sgugger 

Thanks for the input. I changed my pull request up to be more like #17444. Let me know what you think. Thanks!

File: src/transformers/modeling_utils.py
Comment (hackyon) [inline src/transformers/modeling_utils.py:3476]: I took the liberty to add checks just for the start/end of the input_ids to reduce overhead, what do you think of this?

My understanding is that, in most cases, the padding should only show up in the beginning or the end, and that this should catch many of the mistakes.

File: src/transformers/modeling_utils.py
Comment (sgugger) [inline src/transformers/modeling_utils.py:3473]: Let's not add this when we have `warn_once`.

File: src/transformers/modeling_utils.py
Comment (sgugger) [inline src/transformers/modeling_utils.py:3496]: Let's use `warning_once` (or `warn_once` I never remember the name 😅 ) here

**Review by sgugger — COMMENTED** [review]
Thanks for iterating, I think it's better this way. Let's jsut reuse the `warning_once` API we have.

**Comment by hackyon** [comment]
Thanks, I've updated the code accordingly.

**Comment by sgugger** [comment]
@gante could you have a second look here?

**Review by gante — APPROVED** [review]
This is a nice comprehensive warning, I like it! 🔥

**Review by sgugger — APPROVED** [review]

**Comment by gante** [comment]
@hackyon Thank you for the contribution! Would you like to add it to the remaining models? 🤗

**Comment by hackyon** [comment]
Sure, I'll look into it 👍

**Comment by hackyon** [comment]
Thanks @ydshieh for fixing the flaky test!

I was busy in July, but will now add the warning to more models over the next couple of days.

**Pull request merged**
